### PR TITLE
endpoint component enhanced  in trace-search 

### PR DESCRIPTION
**Pull request by horber**
refer to commit notes

### CODE REVIEW

**Comment by wu-sheng** [comment]
Please provide screenshots about what is the latest status.

**Comment by horber** [comment]
The list of endpoints can be obtained through the service key, so it may be more user-friendly to support the drop-down box selection. Otherwise, the search efficiency may be reduced by forgetting which endpoint names exist

**Comment by wu-sheng** [comment]
> The list of endpoints can be obtained through the service key, so it may be more user-friendly to support the drop-down box selection. Otherwise, the search efficiency may be reduced by forgetting which endpoint names exist

This makes sense to me. Just please provide the screenshot to show it first.

**Comment by horber** [comment]
image

**Comment by wu-sheng** [comment]
Make sense. @Fine0830 Could you verify the codes?

**Comment by Fine0830** [comment]
Sure

**Comment by Fine0830** [comment]
I have checked this and have found the following problems

1. The component style of `TraceSearch` is wrong.
2. When we select a service, the logic should only fetch endpoints and instance of this service, rather than fetching services,  endpoints and instance of this service, traces.
3. Set a default value for endpoints name. Should we add `All` in endpoints selector? @wu-sheng 
test

Please fix those. Thank you for your contribution @horber.

**Comment by wu-sheng** [comment]
> When we select a service, the logic should only fetch endpoints and instance of this service, rather than fetching services, endpoints and instance of this service, traces.

Yes, the logic should be as same as the dashboard. Only related instances and endpoints should be shown up. Keep `All` only if no service selected.

> Set a default value for endpoints name. Should we add All in endpoints selector? 

Yes

**Comment by horber** [comment]
Actually, I prefer to search traces by entering key, but I don't feel it easy to add Keyup event ,  I find that the component has been created many times and not destroyed, which will result in multiple binding events and can't be unbound. I wonder if you have found this problem

**Comment by wu-sheng** [comment]
> I prefer to search traces by entering key, but I don't feel it easy to add Keyup event 

I think I have said before, this is not allowed.

**Comment by horber** [comment]
> > I prefer to search traces by entering key, but I don't feel it easy to add Keyup event
> 
> I think I have said before, this is not allowed. In the real product env and large-scale deployment, this will be a disaster.

What I want to express is to search through the shortcut key, such as enter

**Comment by horber** [comment]
However, I found that many components in the project seem to have no reuse mechanism and will be rebuilt in a large number, so that there will be problems with binding shortcuts

**Comment by wu-sheng** [comment]
I think dashboard has that for a while, please take a reference. Actually all logic should share with dashboard selector.

**Comment by Fine0830** [comment]
> However, I found that many components in the project seem to have no reuse mechanism and will be rebuilt in a large number, so that there will be problems with binding shortcuts

Huh, show your code.

**Comment by horber** [comment]
> I have checked this and have found the following problems
> 
> 1. The component style of `TraceSearch` is wrong.
> 2. When we select a service, the logic should only fetch endpoints and instance of this service, rather than fetching services,  endpoints and instance of this service, traces.
> 3. Set a default value for endpoints name. Should we add `All` in endpoints selector? @wu-sheng
>    test
> 
> Please fix those. Thank you for your contribution @horber.

Fixed, please check @Fine0830

**Review by Fine0830 — CHANGES_REQUESTED** [review]
Style problems still exist


JS errors


Here don't need to set endpointName value in the query

File: src/views/components/trace/trace-search.vue [hunk 1]
```diff
@@ -132,16 +141,12 @@ limitations under the License. -->
...
+    private endpoint: Option = { label: 'All', key: '' };
     private traceId: string = localStorage.getItem('traceId') || '';
     private traceState: Option = { label: 'All', key: 'ALL' };
     private tags: string = '';
     private tagsList: string[] = [];
-
     private created() {
-      this.endpointName = this.$route.query.endpointname
-        ? this.$route.query.endpointname.toString()
-        : this.endpointName;
```
Comment (Fine0830) [inline src/views/components/trace/trace-search.vue:144]: After delete this, we aren't able to visit a link with endpointName, link trace id. Should we keep this? @wu-sheng

Comment (wu-sheng) [reply src/views/components/trace/trace-search.vue:144]: I think we only should keep `traceid` as the URI parameter for now.

Comment (wu-sheng) [reply src/views/components/trace/trace-search.vue:144]: @Fine0830 Please confirm, after this change, whether endpoint id is not required in the query? If so, I am willing to remove it from query protocol if the CLI side is fine with that.

Comment (Fine0830) [reply src/views/components/trace/trace-search.vue:144]: Yes, the endpoint id is not required here.

File: src/views/components/trace/trace-search.vue [hunk 2]
```diff
@@ -172,24 +176,40 @@ limitations under the License. -->
...
+        keyword: '',
+        duration: this.durationTime,
+      }).then((data: Array<{ key: string; label: string }>) => {
+        (this.rocketTrace as any).endpoints = data;
+        this.$forceUpdate();
+      });
+    }
+    private chooseInstance(i: any) {
+      this.instance = i;
+      this.getTraceList();
```
Comment (Fine0830) [inline src/views/components/trace/trace-search.vue:203]: There is no need to fetch traces.

File: src/views/components/trace/trace-search.vue [hunk 3]
```diff
@@ -172,24 +176,40 @@ limitations under the License. -->
...
     }
-
     private chooseStatus(i: any) {
       this.traceState = i;
+      this.getTraceList();
```
Comment (Fine0830) [inline src/views/components/trace/trace-search.vue:207]: Same thing as above.

File: src/views/components/trace/trace-search.vue [hunk 4]
```diff
@@ -172,24 +176,40 @@ limitations under the License. -->
...
+    }
+    private chooseEndpoint(i: any) {
+      this.endpoint = i;
+      this.getTraceList();
```
Comment (Fine0830) [inline src/views/components/trace/trace-search.vue:211]: Same thing as above.

**Comment by horber** [comment]
> Style problems still exist
> 
> 
> JS errors
> 
> 
> Here don't need to set endpointName value in the query
> 

fixed

**Review by Fine0830 — CHANGES_REQUESTED** [review]
After changing the endpoint selector, there is no `All`.


There are some endpoints in the selector, when the service is `All`.

File: src/views/components/trace/trace-search.vue [hunk 5]
```diff
@@ -172,24 +170,34 @@ limitations under the License. -->
...
+        (this.rocketTrace as any).endpoints = data;
+        this.$forceUpdate();
+      });
     }
-
     private chooseStatus(i: any) {
       this.traceState = i;
     }
-
+    private chooseEndpoint(i: any) {
```
[the remark this answers is not in the document]
Comment (horber) [reply src/views/components/trace/trace-search.vue:198]: Sure,  before is copied from others

File: src/views/components/trace/trace-search.vue [hunk 6]
```diff
@@ -172,24 +170,34 @@ limitations under the License. -->
...
         return;
       }
       this.GET_INSTANCES({ duration: this.durationTime, serviceId: i.key });
+      (this.rocketTrace as any).endpoints = [];
+      this.GET_ITEM_ENDPOINTS({
+        serviceId: i.key,
+        keyword: '',
+        duration: this.durationTime,
+      }).then((data: Array<{ key: string; label: string }>) => {
+        (this.rocketTrace as any).endpoints = data;
```
[the remark this answers is not in the document]
Comment (horber) [reply src/views/components/trace/trace-search.vue:191]: ok

**Comment by horber** [comment]
all fix, please check @Fine0830

**Review by Fine0830 — APPROVED** [review]
LGTM

**Pull request merged**